### ISSUE TITLE
Foreach – return to C# 4.0 implementation.

### ISSUE THREAD

**Issue by JiriCepelkaFirstLineSoftware** [dotnet/csharplang#3212]
As of C# 5.0 foreach statement was redeclared so now it works like 

```
E e = ((C)(x)).GetEnumerator();
try 
{
	while (e.MoveNext())
	{
		V v = (V)(T)e.Current;
                _embedded-statement_
	}
}
finally { … // Dispose e }
```

While in C# 4.0 that was like

```
E e = ((C)(x)).GetEnumerator();
try {
	V v;
	while (e.MoveNext())
	{
		v = (V)(T)e.Current;
                _embedded-statement_
	}
}
finally { … // Dispose e }
```

As I believe this change was decided because of closing over iteration variable issue when only last value of iteration variable is available for closures.

Far a way as I understand how C# is commonly used it is obviously that user code closes over foreach iteration variable in very little cases while pure enumeration of values has total prevalence. For such a reason allocating new variable on a heap in each iteration is more inconvenient performance hit that with ease fixable issue of closing is inconvenient hit.

Further more such _harmful_ closing over iteration variable perfectly follows well know programming paradigm. Let see proof.

### Proof
Code using foreach statement like

```
foreach (var i in Enumerable.Range(-1, 1))
{
   MakeIGreat(i);
}
```
is perfectly understandable like

```
int i;
foreach (i in Enumerable.Range(-1, 1))
{
   MakeIGreat(i);
}
```

Basically this **inline declaration is** for instance **common** to _out variable declaration_ and _pattern matching_.

### Conclusion

Although this change should did good, it did not. It made precedence for rare case with easy fix while this bug-nobug perfectly follows how declarative statements works over performance making foreach much more worse choose over other loops then before.

So something that should make coding easier made it no easier.

**Comment by HaloFour** [comment]
It made it much easier in the majority of cases.  Yes, a few cases were broken, that was a known and acceptable result.

**Comment by agocke** [comment]
No.

**Comment by JiriCepelkaFirstLineSoftware** [comment]
I am some kind confused today most likely. I interchanged allocation with declaration. Beside different variable scope there is no issue with variable declaring at all.

One thing still persists. Althought some or many users expects variable redeclaration at each loop start variable inline declartion is more declarative expectation.

**Comment by agocke** [comment]
@JiriCepelkaFirstLineSoftware I'm not sure I understand. Are you proposing we alter the behavior of an existing feature, or add a new feature?

**Comment by CyrusNajmabadi** [comment]
This statement doesn't seem true at all:

> Far a way as I understand how C# is commonly used it is obviously that user code closes over foreach iteration variable in very little cases 

The reason we changed the behavior here was because this pattern was happening a lot, and in all (or nearly all) cases the old behavior was not what people expected and it caused problems. 

This was one of the few times an explicit breaking change was made *to the language*. That almost never happens. But it was done precisely here because the original behavior was so understand and confusing for people and the new behavior so much more matched intuition and expectation.

**Comment by JiriCepelkaFirstLineSoftware** [comment]
> @JiriCepelkaFirstLineSoftware I'm not sure I understand. Are you proposing we alter the behavior of an existing feature, or add a new feature?

Now I propose nothing.

**Comment by JiriCepelkaFirstLineSoftware** [comment]
> This statement doesn't seem true at all:
> 
> >…



> This statement doesn't seem true at all:
> 
> > Far a way as I understand how C# is commonly used it is obviously that user code closes over foreach iteration variable in very little cases
> 
> The reason we changed the behavior here was because this pattern was happening a lot, and in all (or nearly all) cases the old behavior was not what people expected and it caused.…

I now these arguments. If you have a look at any foreach you can clearly see that declaration of variable is stated outside the loop execution block thus it is (was) not redeclared in each turn.

That is my understanding of code convention no matter of people senses.

**Comment by CyrusNajmabadi** [comment]
>That is my understanding of code convention no matter of people senses. 

Understood. But we optimized for everyone else's understanding *and* the far more useful and desirable behavior. 

It was intentional and definitely will not be undone. Thanks for the report though! :-)

**Comment by JiriCepelkaFirstLineSoftware** [comment]
Ok. NP. If I did not confused allocation with declaration no issue would be created. Still I do not find better to follow come over exception.
